src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/simple/DPBasedJoinPlanOptimizer.java

```diff
@@ -0,0 +1,164 @@
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpRequest;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpTPAdd;
+import se.liu.ida.hefquin.engine.queryplan.utils.LogicalOpUtils;
+import se.liu.ida.hefquin.engine.queryplan.utils.PhysicalPlanFactory;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizationException;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.QueryOptimizationContext;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting.rules.IdentifyTypeOfRequestUsedForReq;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.utils.PhysicalPlanWithCost;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.utils.PhysicalPlanWithCostUtils;
+import se.liu.ida.hefquin.engine.utils.Pair;
+
+import java.util.*;
+
+public class DPBasedJoinPlanOptimizer extends JoinPlanOptimizerBase {
+
+    protected final QueryOptimizationContext ctxt;
+
+    public DPBasedJoinPlanOptimizer( final QueryOptimizationContext ctxt ) {
+        assert ctxt != null;
+
+        this.ctxt = ctxt;
+    }
+
+    @Override
+    protected EnumerationAlgorithm initializeEnumerationAlgorithm( final List<PhysicalPlan> subplans ) {
+        return new DynamicProgrammingOptimizerImpl(subplans);
+    }
+
+    protected class DynamicProgrammingOptimizerImpl implements EnumerationAlgorithm {
+        protected final List<PhysicalPlan> subplans;
+
+        public DynamicProgrammingOptimizerImpl( final List<PhysicalPlan> subplans ) {
+            this.subplans = subplans;
+        }
+
+        @Override
+        public PhysicalPlan getResultingPlan() throws QueryOptimizationException {
+
+            // Create a data structure that will be used to store the optimal plan for each subset of (sub)plans.
+            final DataStructureForStoringPlansOfSubsets optPlan = new DataStructureForStoringPlansOfSubsets();
+
+            for ( final PhysicalPlan plan: subplans ){
+                optPlan.add( new ArrayList<>(), plan );
+            }
+
+            for ( int num = 2; num < subplans.size(); num ++ ){
+                // Get all subsets with size num.
+                final List<List<PhysicalPlan>> subsets = getSubSet(subplans, num);
+
+                for( final List<PhysicalPlan> plans : subsets ){
+                    final List<PhysicalPlan> candidatePlans = new ArrayList<>();
+
+                    // Split the current set of subplans into two subsets, and create candidate plans with join for each of the combinations.
+                    final List<Pair<List<PhysicalPlan>, List<PhysicalPlan>>> candidatePairs = splitIntoSubSets(plans);
+                    for ( Pair p: candidatePairs ){
+                        final PhysicalPlan plan_left = optPlan.get((List<PhysicalPlan>) p.object1);
+                        final PhysicalPlan plan_right = optPlan.get((List<PhysicalPlan>) p.object2);
+
+                        candidatePlans.add( PhysicalPlanFactory.createPlanWithJoin( plan_left,  plan_right) );
```

Comment:
Before doing that, include a check to verify that `plan_left` is not `null` and `plan_right` is not `null`. In principle, they should not, but better verify to avoid weird bugs. If they are `null`, for whatever reason, throw an `IllegalStateException`.